[Clang] Add  constexpr support for AVX512 permutexvar intrinsics

chaitanyav opened the pull request:
Resolves: #167476

chaitanyav committed:
[Clang] VectorExprEvaluator::VisitCallExpr / InterpretBuiltin - Allow AVX512 permutexvar intrinsics to be used in constexpr

Resolves: #167476

RKSimon commented on clang/lib/AST/ByteCode/InterpBuiltin.cpp:
```diff
@@ -4414,6 +4414,50 @@ bool InterpretBuiltin(InterpState &S, CodePtr OpPC, const CallExpr *Call,
+        S, OpPC, Call, [](unsigned DstIdx, unsigned ShuffleMask) {
+          int Offset = ShuffleMask & 0x7;
+          unsigned SrcIdx = 0;
+          return std::pair<unsigned, int>{SrcIdx, Offset};
```
Don't bother with SrcIdx definitions:
```suggestion
          return std::pair<unsigned, int>{0, Offset};
```

RKSimon commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -13551,6 +13551,70 @@ bool VectorExprEvaluator::VisitCallExpr(const CallExpr *E) {
+                            [](unsigned DstIdx, unsigned ShuffleMask) {
+                              int Offset = ShuffleMask & 0x7;
+                              unsigned SrcIdx = 0;
+                              return std::pair<unsigned, int>{SrcIdx, Offset};
```
Sane for all these
```suggestion
                              return std::pair<unsigned, int>{0, Offset};
```

chaitanyav committed:
Address code review comments

- Group permvarsi256/permvarsf256 with other AVX2 constexpr builtins
- Remove unnecessary SrcIdx variable and use zero directly in pair
  construction

RKSimon reviewed (approved):
LGTM - cheers

RKSimon reviewed (changes requested):
sorry - you missed the avx2 tests

chaitanyav committed:
Add tests for permutevar8x32, permutevar8x32_epi32

RKSimon reviewed (approved):
LGTM - cheers

RKSimon committed:
Merge branch 'main' into issue_167476
